[MRG] Add Deprecation Warning to address making "batch" the default learning method in LDA in 0.19 release

doshyt opened the pull request:
#### Reference Issue

Fixes: #6997 
#### What does this implement/fix? Explain your changes.

This fix sets "batch" as the default "learning_method" in Latent Dirichlet Allocation.
Also, it fixes a few tests which used the default method and were explicitly said to perform over "online" learning method of LDA. 
Finally, the line stating that "online" is faster than "batch" was removed from the documentation.
#### Any other comments?

What could be better update for the DOC? Do we have a reasoning why "batch" performs better?

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

doshyt committed:
update default LDA method to batch

doshyt committed:
update tests to match default LDA learning scheme

doshyt committed:
DOC trim information on faster learning scheme with LDA

agramfort commented on sklearn/decomposition/online_lda.py:
```diff
@@ -246,7 +246,7 @@ class LatentDirichletAllocation(BaseEstimator, TransformerMixin):
 
     def __init__(self, n_topics=10, doc_topic_prior=None,
-                 topic_word_prior=None, learning_method='online',
+                 topic_word_prior=None, learning_method='batch',
```
this an API change. You cannot do this without a proper deprecation cycle. If it's what "people" want

amueller commented:
From my experience with the code and from talking to @hannawallach this is definitely what we want. But it does need a deprecation. so you need to change the learning_method to ``None``, then check if it was ``None`` in ``fit`` and raise a deprecation warning that the default will change in ``0.19`` (right? I'm bad with numbers).

amueller commented:
See http://scikit-learn.org/dev/developers/contributing.html#deprecation

hannawallach commented:
Yup, I'd set the default to batch (which, IME, generally works better than stochastic) and then let people use stochastic if they have a lot of data.

doshyt commented:
@amueller I would need just a small clarification regarding the deprecation in the context of the referenced issue. Did I get it right that we want to switch to the `batch` learning method in the next release and keep the things as they are now (and as you described in your comment). At the same time, we produce a deprecation warning if the learning method was not specified.

So, currently all what we want is to add the warning and do not change the default behavior yet? Or did I misunderstood and I have to implement both steps at the same time - add warning AND change the default to `batch`?

Sorry if this seems obvious, I'm trying my best to smoothly catch up with the practices of scikit-learn.

amueller commented:
It's not obvious at all, the title of the issue is pretty misleading in this regard. But you are right in your understanding, we don't change anything for now, but just warn and tell people it'll change in the future.

doshyt committed:
Add deprecation warning and revert learning_method back to the default

After the discussion, it was decided that a proper deprecation cycle
for the parameter is required. Therefore, in this PR only the warning is
added. The default learning_method will be changed to 'batch' in 0.19.

doshyt committed:
Revert back tests to the default LDA

doshyt commented:
@amueller I pushed my fixes. By the way, shall I keep an eye on this issue to be changed as promised in 0.19 or is there any tracking system for something that is going to be changed?

jrfiedler commented:
Since the deprecation is happening in 0.18, shouldn't the change occur in 0.20? 

Also (maybe this is nitpicky?) the message should say the version where deprecation happened. See the examples in the section @amueller linked to: http://scikit-learn.org/dev/developers/contributing.html#deprecation.

nelson-liu commented:
+1 on @jrfielder 's comments

Commit:
DOC deprecate in 0.20 instead of 0.19

doshyt commented:
@jrfiedler @nelson-liu I agree with comments, pushed my modifications

amueller commented:
sorry about the one-off error, that's why I wanted to have that in the docs. There is an automatic tacking system, which is me doing a ``git grep`` before the release (don't tell anyone).

amueller commented:
@ogrisel @jnothman we should really put this into 0.18. I thought it was merged. Sorry about that. This is important.

jnothman commented:
+1. Do it

hannawallach commented:
+1 from me too.

On Sat, Sep 17, 2016 at 9:00 AM, Joel Nothman <notifications@github.com>
wrote:

> +1. Do it
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/6999#issuecomment-247768140>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA1T7u6OZC9yTAyf3PNtqtXJpjM9th9eks5qq-R2gaJpZM4JN2Xr>
> .
>



-- 
hanna wallach
http://dirichlet.net/
